pilot/pkg/networking/core/v1alpha3/configgen.go

```diff
@@ -21,14 +21,17 @@ import (
 
 	"istio.io/istio/pilot/pkg/model"
 	"istio.io/istio/pilot/pkg/networking/plugin"
+	"istio.io/istio/pilot/pkg/networking/util"
 )
 
 type ConfigGeneratorImpl struct {
 	// List of plugins that modify code generated by this config generator
 	Plugins []plugin.Plugin
-	// List of outbound clusters keyed by configNamespace. For use by gateways only.
+	// gatewayCDSMutex protects the precomputed outbound clusters for gateways
+	gatewayCDSMutex sync.RWMutex
+	// List of outbound clusters keyed by configNamespace & locality. For use by gateways only.
 	// Must be rebuilt for each push epoch
-	PrecomputedOutboundClustersForGateways map[string][]*xdsapi.Cluster
+	PrecomputedOutboundClustersForGateways map[string]map[string][]*xdsapi.Cluster
```

Comment:
Somehere we must have some documentation for people to understand what is going on.
Do we have a design doc or readme or file with more comments explaining the design ? Some example generated configs would help as well, as well as a local integration test with a realistic locality config that can save the output to file ( see the none tests and golden for example). 

Would it be more efficient to encode the locality in the cluster name ? 
Based on the code - I have little idea of what to expect, since there are no comments or tests.

It seems the change still has an impact for non-locality based use cases - mostly in this struct. From what I see it seems safe - but given the lack of stability/stress/perf tests in this area any problem will probably be found by customers... Current perf tests are very limited. Maybe we can do before/after tests.

Reply:
This is simply an index per namespace and per locality.
 namespace:
   locality:
     list of clusters

Right now, there is only one locality in each namespace (the empty /// locality that we currently use). 

At runtime, we lookup the map based on the calling proxy's namespace first and then by the calling proxy's locality next. Currently, the locality piece is set to /// and we clone the list of clusters, and adjust the locality setting based on the proxy's locality. As I stated in earlier comment, knowing the list of all proxy localities is not possible at the moment. So the cache is effectively the same as the earlier cache. I will add comments to that effect

We cannot encode the locality in the cluster name (locality can be nil, or

Reply:
The cache was added as an early optimization - and it has many limitations and problems, in particular when we do smarter assignment (like locality-based). 

We now have the 'shards' cache, which has the info for each service (broken by clusters for fast update). I don't think we need the cluster cache for 'advanced' endpoint selection.

Most clusters have 2-3 endpoints - it is wasteful to create a map inside a map to cache the list, and will make removal of the cache harder. 

Can we avoid this - and just select from the shards ? Even for large services, with 1000s of endpoints - we'll want to subset (not send 1000 to everyone), and once we have health or load the logic will even more interesting. 

I suppose we can live with an extra map in 1.1...